atorch/examples/ds_3d_parallel/design.md

```diff
@@ -0,0 +1,844 @@
+                num_tokentypes=0,
+                parallel_output=True
+            )
+        else:
+            model = GPTModel(
+                num_tokentypes=0,
+                parallel_output=True,
+                pre_process=pre_process,
+                post_process=post_process
+            )
+
+    class GPTModelPipe(PipelineModule,MegatronModule):
+        def __init__(self, ...):
+            ...
+            # Embedding layer
+            self.specs.append(TiedLayerSpec('embed',
+                                            EmbeddingPipe,
+                                            args.hidden_size,
+                                            args.padded_vocab_size,
+                                            args.max_position_embeddings,
+                                            args.hidden_dropout,
+                                            init_method=init_method,
+                                            num_tokentypes=num_tokentypes,
+                                            tied_weight_attr='word_embeddings_weight'))
+
+            for layer_idx in range(args.num_layers):
+                self.specs.append(
+                    LayerSpec(ParallelTransformerLayerPipe,
+                        init_method=init_method,
+                        output_layer_init_method=scaled_init_method_normal(args.init_method_std,
+                                                                        args.num_layers),
+                        layer_number=layer_idx,
+                        self_attn_mask_type=AttnMaskType.causal))
+                        
+            def _logits_helper(embedding, lm_output):
+                """A wrapper to massage inputs/outputs from pipeline. """
+                return parallel_lm_logits(
+                    lm_output,
+                    embedding.word_embeddings_weight,
+                    self.parallel_output)
+
+            self.specs.append(
+                TiedLayerSpec('embed',
+                            EmbeddingPipe,
+                            args.hidden_size,
+                            args.padded_vocab_size,
+                            args.max_position_embeddings,
+                            args.hidden_dropout,
+                            init_method=init_method,
+                            num_tokentypes=num_tokentypes,
+                            forward_fn=_logits_helper,
+                            tied_weight_attr='word_embeddings_weight')
+            )
+    ```
+
+    </details>
+
+
+
+- The HuggingFace Transformers library provides native torch implementations of a large number of NLP models, which are constantly emerging. Megatron Deepspeed hybrid parallel support is limited.
+- The goal: to reuse deepspeed/megatron capabilities to provide a hybrid parallelism suitable for a wide range of native Torch implementations, and integrate into Atorch auto_accelerate optimization using a low-cost hybrid parallelism scheme.
+# User Interface
```

Comment:
Main purpose of this doc is explaining how to use ds_3d_parallel by example. 
So, many of the contexts can be omitted below.

Reply:
`tutorial.md` explains how to use ds_3d_parallel by example, `design.md` mainly describes the implementation.

Reply:
[remove design.md; merge intro to tutorial.md](https://github.com/intelligent-machine-learning/dlrover/pull/783/commits/a5fdcca2b03f36fd21017f90b6313b1346187132)